ambry-router/src/main/java/com/github/ambry/router/DeleteOperation.java

```diff
@@ -321,21 +309,17 @@ private void checkAndMaybeComplete() {
         operationException.set(
             new RouterException("DeleteOperation failed because of BlobNotFound", RouterErrorCode.BlobDoesNotExist));
       }
+      if(quotaChargeCallback != null) {
```

Comment:
I think there is an issue here. `checkAndMaybeComplete` is intended as an idempotent method that can be called multiple times. For example, let's look at the `handleResponse`  call. This is called for every response that comes back from a server. I think it may be possible for a response to come back such that the success threshold for the operation comes back, and this code block will be executed. Then, the 3rd successful response may come back and this would be called again.